lib/address.py

```diff
@@ -311,21 +312,49 @@ def validate(cls, pubkey):
 
     @cachedproperty
     def address(self):
-        '''Convert to an Address object.'''
+        '''Convert to a P2PKH Address object.'''
         return Address(hash160(self.pubkey), Address.ADDR_P2PKH)
 
     def is_compressed(self):
         '''Returns True if the pubkey is compressed.'''
         return len(self.pubkey) == 33
 
     def to_ui_string(self):
-        '''Convert to a hexadecimal string.'''
+        '''Convert to a hexadecimal string.
+
+        (do not change this -- many places rely on this to be hexadecimal)
+        '''
         return self.pubkey.hex()
 
     def to_storage_string(self):
         '''Convert to a hexadecimal string for storage.'''
         return self.pubkey.hex()
 
+    def to_P2PKH_script(self):
+        '''Return a P2PKH script.'''
+        return self.address.to_script()
+
+    def __str__(self):
+        return self.to_ui_string()
+
+    def __repr__(self):
+        return '<PubKey {}>'.format(self.__str__())
+
+
+class P2PKOutput(namedtuple("P2PKTuple", "pubkey")):
+    """ Output type for pay-to-pubkey scripts (<pubkey> OP_CHECKSIG) """
+
+    @classmethod
+    def from_pubkey(cls, pubkey):
+        # expects bytes
+        PublicKey.validate(pubkey)
+        return cls(bytes(pubkey))
+
+    def to_ui_string(self):
+        '''Convert to a human-friendly string.
+        '''
+        return 'P2PK(%s)'%(self.pubkey.hex())
```

Comment:
Are we sure this is safe? Lol.. granted it can't lead to signing errors.. but you'd be surprised with this crazy codebase. Ha ha.

Reply:
The idea is that `to_ui_string` here is only going to be called in same places where it's called for ScriptOutputs, so that it will be totally ok. :-)

But, I must continue digging. :-)

Reply:
You are literally adding a P2PKOutput to `inputs` in this very PR -- which will end up causing the same thing that broke in 4.0.4 release... because code relies on `to_ui_string` everywhere.. :)